### ISSUE TITLE
 backward_loss transpose() meaning in MultipleNegativesSymmetricRankingLoss 

### ISSUE THREAD

**Issue by sogmgm** [UKPLab/sentence-transformers#2522]
Hello, I just want to say a big thank you for sentence-transformer. 

I have  a question  for backward_loss  in MultipleNegativesSymmetricRankingLoss 
according to code 
```
import torch
from torch import nn, Tensor
from typing import Iterable, Dict
import sys
sys.path.append("/content/drive/MyDrive/논문/sentence_transformers")
from sentence_transformers import SentenceTransformer
from sentence_transformers import util

class MultipleNegativesSymmetricRankingLoss(nn.Module):
    def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_fct=util.cos_sim):
        """
        This loss is an adaptation of MultipleNegativesRankingLoss. MultipleNegativesRankingLoss computes the following loss:
        For a given anchor and a list of candidates, find the positive candidate.

        In MultipleNegativesSymmetricRankingLoss, we add another loss term: Given the positive and a list of all anchors,
        find the correct (matching) anchor.

        For the example of question-answering: You have (question, answer)-pairs. MultipleNegativesRankingLoss just computes
        the loss to find the answer for a given question. MultipleNegativesSymmetricRankingLoss additionally computes the
        loss to find the question for a given answer.

        Note: If you pass triplets, the negative entry will be ignored. A anchor is just searched for the positive.

        :param model: SentenceTransformer model
        :param scale: Output of similarity function is multiplied by scale value
        :param similarity_fct: similarity function between sentence embeddings. By default, cos_sim. Can also be set to dot product (and then set scale to 1)

        Requirements:
            1. (anchor, positive) pairs

        Relations:
            - Like :class:`MultipleNegativesRankingLoss`, but with an additional loss term.

        Inputs:
            +---------------------------------------+--------+
            | Texts                                 | Labels |
            +=======================================+========+
            | (anchor, positive) pairs              | none   |
            +---------------------------------------+--------+

        Example:
            ::

                from sentence_transformers import SentenceTransformer, losses, InputExample
                from torch.utils.data import DataLoader

                model = SentenceTransformer('distilbert-base-uncased')
                train_examples = [
                    InputExample(texts=['Anchor 1', 'Positive 1']),
                    InputExample(texts=['Anchor 2', 'Positive 2']),
                ]
                train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=32)
                train_loss = losses.MultipleNegativesSymmetricRankingLoss(model=model)
                model.fit(
                    [(train_dataloader, train_loss)],
                    epochs=10,
                )
        """
        super(MultipleNegativesSymmetricRankingLoss, self).__init__()
        self.model = model
        self.scale = scale
        self.similarity_fct = similarity_fct
        self.cross_entropy_loss = nn.CrossEntropyLoss()

    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
        reps = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]
        anchor = reps[0]
        candidates = torch.cat(reps[1:])

        scores = self.similarity_fct(anchor, candidates) * self.scale
        labels = torch.tensor(
            range(len(scores)), dtype=torch.long, device=scores.device
        )  # Example a[i] should match with b[i]

        anchor_positive_scores = scores[:, 0 : len(reps[1])]
        forward_loss = self.cross_entropy_loss(scores, labels)
        backward_loss = self.cross_entropy_loss(**anchor_positive_scores.transpose(0, 1)**, labels)
        return (forward_loss + backward_loss) / 2

    def get_config_dict(self):
        return {"scale": self.scale, "similarity_fct": self.similarity_fct.__name__}
```

 backward_loss = self.cross_entropy_loss(**anchor_positive_scores.transpose(0, 1)**, labels)

anchor_positive_scores is  a score  matrix( bs x bs )  and  numbers in matrix  means cos_sim (each anchor i , positive i )

so i guess anchor_positive_scores  is symmetric, but i have  question why this  code uses `transpose(1,0)`

 regards

**Comment by tomaarsen** [comment]
Hello!

`anchor_positive_scores` is indeed a square matrix with batch_size x batch_size as the shape, but matrix is not symmetric: e.g. `anchor_positive_scores[2][1]` != `anchor_positive_scores[1][2]`, because `anchor_positive_scores[2][1]` is the score for the anchor at index 2 with the positive at index 1 and `anchor_positive_scores[1][2]` is the anchor at index 1 with the positive at index 2.

Also, in my experiment it seems like `anchor_positive_scores` is actually just a copy of `scores` (i.e. if you call the loss with just an anchor and a positive like you're supposed to). So, then:
1. **self.cross_entropy_loss(scores, labels)**: This means: Given the anchor, how close is it to the correct positive, while being far away from all the `batch_size-1` unrelated positives?
2. **self.cross_entropy_loss(scores.transpose(0, 1), labels)**: This means: Given the **positive**, how close is it to the correct **anchor**, while being far away from all the `batch_size-1` unrelated anchors?

That is unless I am mistaken, but I think that's right. Note that the anchor & positive sentences can be very different, e.g. search query vs search document, so it's uncommon to train such that you can also detect related queries given a document (which is what this loss allows).

Hope that helps a bit.

- Tom Aarsen

**Comment by sogm1** [comment]
@tomaarsen 


I got it  `anchor_positive_scores[2][1] != anchor_positive_scores[1][2]`  mean . I misunderstood that. 

Also I make sense your comment.  great!! 
`self.cross_entropy_loss(scores.transpose(0, 1), labels): This means: Given the positive, how close is it to the correct anchor, while being far away from all the batch_size-1 unrelated anchor
`
Thank U so much. 
Regrads
